registry/registry.go

```diff
@@ -741,20 +833,247 @@ func downloadVersion(opts *VersionOptions) (ver *Version, attachments []*kivik.A
+			for _, shot := range locale.Screenshots {
+				shot = path.Join("/", shot)
+				if !stringInArray(shot, screenshotPaths) {
+					screenshotPaths = append(screenshotPaths, shot)
+				}
+			}
+		}
+	}
+
+	// Re-reading tarball content for assets
+	if len(screenshotPaths) > 0 || iconPath != "" || partnershipIconPath != "" {
+		_, err := buf.Seek(0, io.SeekStart)
+		if err != nil {
+			return nil, err
+		}
+		tr, err := tarReader(buf, tarball.ContentType)
+		if err != nil {
+			err = errshttp.NewError(http.StatusUnprocessableEntity,
+				"Could not reach version on specified url %s: %s", url, err)
+			return nil, err
+		}
+
+		for {
+			var hdr *tar.Header
+			hdr, err = tr.Next()
+			if err == io.EOF {
+				break
+			}
+			if err == io.ErrUnexpectedEOF {
+				err = errshttp.NewError(http.StatusUnprocessableEntity,
+					"Could not reach version on specified url %s: file is too big %s", url, err)
+				return nil, err
+			}
+			if err != nil {
+				err = errshttp.NewError(http.StatusUnprocessableEntity,
+					"Could not reach version on specified url %s: %s", url, err)
+				return nil, err
+			}
+
+			if hdr.Typeflag != tar.TypeReg && hdr.Typeflag != tar.TypeDir {
+				continue
+			}
+
+			name := path.Join("/", hdr.Name)
+			if tarball.TarPrefix != "" {
+				name = path.Join("/", strings.TrimPrefix(name, tarball.TarPrefix))
+			}
+			if name == "/" {
+				continue
+			}
+
+			isIcon := iconPath != "" && name == iconPath
+			isPartnershipIcon := partnershipIconPath != "" && name == partnershipIconPath
+
+			isShot := !isIcon && stringInArray(name, screenshotPaths)
+			if !isIcon && !isPartnershipIcon && !isShot {
+				continue
+			}
+
+			var data []byte
+			data, err = ioutil.ReadAll(tr)
+			if err != nil {
+				err = errshttp.NewError(http.StatusUnprocessableEntity,
+					"Could not reach version on specified url %s: %s", url, err)
+				return nil, err
+			}
+			var filename string
+			if isIcon {
+				filename = "icon"
+			} else if isShot {
+				filename = name
+			} else if isPartnershipIcon {
+				filename = "partnership_icon"
+			} else {
+				panic("unreachable")
+			}
+			mime := magic.MIMEType(name, data)
+			body := ioutil.NopCloser(bytes.NewReader(data))
+			attachments = append(attachments, &kivik.Attachment{
+				Content:     body,
+				Size:        int64(len(data)),
+				Filename:    filename,
+				ContentType: mime,
+			})
+		}
+	}
+	return attachments, nil
+}
+
+// SaveTarball saves tarball to swift
+func SaveTarball(space, filepath, contentType string, fileContent io.Reader) error {
+	// Saving the tar to Swift
+	conf, err := config.GetConfig()
+	if err != nil {
+		return err
+	}
+	sc := conf.SwiftConnection
+
+	f, err := sc.ObjectCreate(space, filepath, false, "", contentType, nil)
+	if err != nil {
+		return err
+	}
+
+	defer f.Close()
+
+	_, err = io.Copy(f, fileContent)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// ReadTarballVersion reads the content of the version tarball which has been
+// downloaded
```

Comment:
It looks like the function is doing a lot more than what is said in this comment